rclpy/test/test_clock.py

```diff
@@ -211,6 +216,137 @@ def test_triggered_clock_change_callbacks(self):
         handler3.unregister()
 
 
+@pytest.fixture()
+def default_context():
+    rclpy.init()
+    yield get_default_context()
+    rclpy.shutdown()
+
+
+@pytest.fixture()
+def non_default_context():
+    context = Context()
+    context.init()
+    yield context
+    context.try_shutdown()
+
+
+def test_sleep_until_mismatched_clock_type(default_context):
+    clock = Clock(clock_type=ClockType.SYSTEM_TIME)
+    with pytest.raises(ValueError, match='.*clock type does not match.*'):
+        clock.sleep_until(Time(clock_type=ClockType.STEADY_TIME))
+
+
+def test_sleep_until_non_default_context(non_default_context):
+    clock = Clock()
+    assert clock.sleep_until(clock.now() + Duration(seconds=0.1), context=non_default_context)
+
+
+def test_sleep_until_invalid_context():
+    clock = Clock()
+    with pytest.raises(NotInitializedException):
+        clock.sleep_until(clock.now() + Duration(seconds=0.1), context=Context())
+
+
+@pytest.mark.parametrize(
+    'clock_type', (ClockType.SYSTEM_TIME, ClockType.STEADY_TIME, ClockType.ROS_TIME))
+def test_sleep_until_basic(default_context, clock_type):
+    clock = Clock(clock_type=clock_type)
+    sleep_duration = Duration(seconds=0.1)
+    start = clock.now()
+    assert clock.sleep_until(clock.now() + sleep_duration)
+    stop = clock.now()
+    assert stop - start >= sleep_duration
+
+
+@pytest.mark.parametrize(
+    'clock_type', (ClockType.SYSTEM_TIME, ClockType.STEADY_TIME, ClockType.ROS_TIME))
+def test_sleep_until_time_in_past(default_context, clock_type):
+    clock = Clock(clock_type=clock_type)
+    sleep_duration = Duration(seconds=-1)
+    start = clock.now()
+    assert clock.sleep_until(clock.now() + sleep_duration)
+    stop = clock.now()
+    a_small_amount_of_time = Duration(seconds=0.01)
+    assert stop - start < a_small_amount_of_time
+
+
+@pytest.mark.parametrize('ros_time_enabled', (True, False))
+def test_sleep_until_ros_time_toggled(default_context, ros_time_enabled):
+    clock = ROSClock()
+    clock._set_ros_time_is_active(not ros_time_enabled)
+
+    retval = None
+
+    def run():
+        nonlocal retval
+        retval = clock.sleep_until(clock.now() + Duration(seconds=999999))
```

Comment:
I'm a bit worried about this very long timeout.  In the case that a bug is introduced, this will wait for a *long* time.  Maybe we should just reduce this to 10 seconds or something?

Reply:
I can shorten it. Right now if a bug is introduced, the max it will wait until ctest kills it is 120 seconds

https://github.com/ros2/rclpy/blob/3af017844a214a838a27891710cc9bee52331842/rclpy/CMakeLists.txt#L229

Reply:
Shortened to 10 seconds in 0530656493a2de30baace5f2cde20e73527092bf

Reply:
Oh right, I forgot that `rclpy` uses `ctest` and not `pytest`.  Still, a shorter timeout in the case of a bug is a good thing, I think.  The timeout of 10 seconds seems reasonable to me, thanks.

Reply:
The 99999 timeout was also indirectly testing that the methods returned sooner than the duration of time they were asked to sleep. In 8615b7b I added explicit checks that they return in a small amount of time.